wfe2/wfe.go

```diff
@@ -2353,14 +2354,111 @@ func (wfe *WebFrontEndImpl) FinalizeOrder(ctx context.Context, logEvent *web.Req
+// certificate. It is implemented by both the deprecatedCertID and certID
+// structs.
+//
+// TODO(#7183): Remove this.
+type ariCertID interface {
+	Serial() string
+}
+
+// deprecatedCertID matches the ASN.1 structure of the CertID sequence defined
+// by RFC6960. Its fields are exported so that they can be unmarshaled using the
+// asn1 package.
+//
+// TODO(#7183): Remove this.
+type deprecatedCertID struct {
 	HashAlgorithm  pkix.AlgorithmIdentifier
 	IssuerNameHash []byte
 	IssuerKeyHash  []byte
 	SerialNumber   *big.Int
 }
 
+func (c deprecatedCertID) Serial() string {
+	return core.SerialToString(c.SerialNumber)
+}
+
+// newDeprecatedCertID parses a unique identifier (certID) for a certificate as
+// per the ACME protocol's "renewalInfo" resource, as specified in draft-ietf-
+// acme-ari-02. For more details see:
+// https://datatracker.ietf.org/doc/html/draft-ietf-acme-ari-01#section-4.1.
+//
+// TODO(#7183): Remove this.
+func newDeprecatedCertID(path string) (ariCertID, error) {
+	der, err := base64.RawURLEncoding.DecodeString(path)
+	if err != nil {
+		return deprecatedCertID{}, err
+	}
+
+	var id deprecatedCertID
+	rest, err := asn1.Unmarshal(der, &id)
+	if err != nil || len(rest) != 0 {
+		return deprecatedCertID{}, err
+	}
+
+	// Verify that the hash algorithm is SHA-256, so people don't use SHA-1 here.
+	if !id.HashAlgorithm.Algorithm.Equal(asn1.ObjectIdentifier{2, 16, 840, 1, 101, 3, 4, 2, 1}) {
+		// wfe.sendError(response, logEvent, probs.Malformed("Request used hash algorithm other than SHA-256"), err)
+		return deprecatedCertID{}, errors.New("Request used hash algorithm other than SHA-256")
+	}
+	return id, nil
+}
+
+// certID represents a unique identifier (certID) for a certificate as per the
+// ACME protocol's "renewalInfo" resource, as specified in draft-ietf-acme-ari-
+// 02. The certID is a composite string derived from the base64url-encoded
+// keyIdentifier of the certificate's Authority Key Identifier (AKI) and the
+// base64url-encoded serial number of the certificate, separated by a period.
+// For more details see:
+// https://datatracker.ietf.org/doc/html/draft-ietf-acme-ari-02#section-4.1.
+type certID struct {
+	keyIdentifier []byte
+	serialNumber  *big.Int
+}
+
+func (c certID) Serial() string {
+	return core.SerialToString(c.serialNumber)
+}
+
+// newCertID parses a unique identifier (certID) as specified in
+// draft-ietf-acme-ari-02. It takes the composite string as input returns a
+// certID struct with the keyIdentifier and serialNumber extracted and decoded.
+// For more details see:
+// https://datatracker.ietf.org/doc/html/draft-ietf-acme-ari-02#section-4.1.
+func newCertID(path string, issuerCertificates map[issuance.IssuerNameID]*issuance.Certificate) (ariCertID, *probs.ProblemDetails, error) {
+	parts := strings.Split(path, ".")
+	if len(parts) != 2 || parts[0] == "" || parts[1] == "" {
+		return certID{}, probs.Malformed("Path was not a composite CertID"), nil
```

Comment:
The term "composite CertID" doesn't show up in the RFC so we shouldn't use it. We can just say "invalid path".